### PR TITLE
norm implementation

### PR DESCRIPTION
**Pull request by mtar**
## Description

Rewrite of norm to support more types of norms.

Issue/s resolved: #836 #845 #847 

## Changes proposed:
- add norm, vector_norm, matrix_norm
- fix bug reduce_op with axis and keepdim
- dot returns a 0-dim array instead of builtin float
- fix bug in max/min (_reduce_op) on distributed arrays.

## Type of change

- New feature
- Bug fix

## Due Diligence

- [x] All split configurations tested
- [x] Multiple dtypes tested in relevant functions
- [x] Documentation updated (if needed)
- [x] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

**Comment by mtar** [comment]
rerun tests

File: heat/core/_operations.py [hunk 1]
```diff
@@ -439,7 +442,7 @@ def __reduce_op(
             balanced = True
             if x.comm.is_distributed():
                 x.comm.Allreduce(MPI.IN_PLACE, partial, reduction_op)
-        elif axis is not None:
+        elif axis is not None and not keepdim:
```
Comment (coquelin77) [inline heat/core/_operations.py:445]: good catch

File: heat/core/linalg/basics.py [hunk 2]
```diff
@@ -752,24 +770,277 @@ def matmul(a: DNDarray, b: DNDarray, allow_resplit: bool = False) -> DNDarray:
...
+    ord    norm for matrices
+    =====  ============================
+    None   Frobenius norm
+    'fro'  Frobenius norm
+    'nuc'  nuclear norm
+    inf    max(sum(abs(x), axis=1))
+    -inf   min(sum(abs(x), axis=1))
+    1      max(sum(abs(x), axis=0))
+    -1     min(sum(abs(x), axis=0))
+    =====  ============================
```
Comment (coquelin77) [inline heat/core/linalg/basics.py:814]: did you check this in the compiled docs? i think that this might make things messy

**Review by coquelin77 — CHANGES_REQUESTED** [review]

File: heat/core/linalg/basics.py (hunk 2 above)
Comment (mtar) [reply heat/core/linalg/basics.py:814]: It works as expected locally

image

**Review by coquelin77 — APPROVED** [review]

**Pull request merged**